### DISCUSSION TITLE
Aperture Stops

### DISCUSSION THREAD

**Discussion by cbhirsch** [mjhoptics/ray-optics#116] [Q&A] [answered]
Here's some example code that I'm using:
```
%matplotlib inline
# initialization
from rayoptics.environment import *

root_pth = Path(rayoptics.__file__).resolve().parent

app = AppManager(None)
opm = app.model = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']

sm.gaps[0].thi = 1e10
sm.list_model()

osp.pupil.key, osp.pupil.value

sm.do_apertures = False
osp.pupil.value=22
opm.update_model()

opm.add_from_file(root_pth/"codev/tests/CODV_47737.seq", t=8.12)
opm.add_from_file(root_pth/"codev/tests/CODV_47663.seq", t=2.01)
opm.update_model()

layout_plt0 = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                         do_draw_rays=True, do_paraxial_layout=False).plot()

```

So I'm following the example "AddFromFile.ipynb" example for the most part this is working for me and I've figured out how to add my own lenses into the system.  Now I'm trying to figure out how I would go about adding a stop to the system and getting the model to generate a chief and marginal ray that I could then define the NA around.

1.  Is it possible to add an aperture stop into the system?
2. How would I go about having it find the chief and marginal rays? 

![example](https://user-images.githubusercontent.com/91820473/232923097-c0514664-4400-4602-8c70-6cf23018d07c.png)

**Comment by mjhoptics** [comment]
Hello @cbhirsch,

The answer is yes for both questions. Here are some details.

You can designate a specific interface to be the stop surface. For example, if you want the first surface of the second doublet to be the stop, directly set the `stop_surface` attribute in the `seq_model`.

`sm.stop_surface = 4`

You can set the stop on the current surface, e.g. when you are building up a model, using

`sm.set_stop()`

If the limiting aperture isn't on a lens/mirror surface, you can add a planar dummy surface in the appropriate airspace and set the stop to that.

To add field points to your model, create and populate a Field of View model (part of the optical specifications). You can specify fields by height or angle, in either the object space or image space. This example uses angle in object space, where the maximum field is at 10.0 deg off-axis. Three fields are created using the `set_from_list()` method.
```
fov = osp['fov']
fov.key = 'field', 'object', 'angle'
fov.value = 10.0   # max field angle of 10.0 degrees
fov.is_relative = True
fov.set_from_list([0, 0.707, 1])
opm.update_model()
```
Alternatively, specifying points on the image plane is often useful. The following key/value pair would be used.
```
fov.key = 'field', 'image', 'height'
fov.value = 21.6   # max image height
```

You can set the image plane to be at the paraxial focus using a convenience method in the parax_model.
```
sm.gaps[-1].thi += pm.thi_ht_solve(pm.ax, -1, 0)
opm.update_model()
```

Feel free to ask more questions on this. Regards, Mike

**Reply by cbhirsch** [reply]
So I tried to switch things up a bit here but still using the aperture stop I'm modeling this simple system and I was using the triplet example.  So below is a problem from Introduction to lens design from J Geary.  I've attached my code for modeling this problem as well.

![Screenshot 2023-04-20 103405](https://user-images.githubusercontent.com/91820473/233432099-ee4a1491-08ca-45ea-882c-c160649f2f10.jpg)

```
isdark = False

# use standard rayoptics environment
from rayoptics.environment import *

# util functions
from rayoptics.util.misc_math import normalize

#Creating a new, empty, model
opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']

#units
opm.system_spec.dimensions = 'mm'

#Specify aperture, field, and wavelengths
osp['pupil'] = PupilSpec(osp, key=['image', 'f/#'], value=10)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=5, flds=[0. , 1.], is_relative=True)
osp['wvls'] = WvlSpec([('F', 0.5), (587.5618, 1.0), ('C', 0.5)], ref_wl=1)
sm.stop_surface = 40

opm.radius_mode = True

sm.gaps[0].thi=1e10

sm.set_stop()
sm.add_surface([206, 1, 'N-BK7', 'Schott'])
sm.add_surface([206, 5.86])

opm.update_model()
sm.list_model()
```

I'm getting the following error output:

`
line 65, in __init__ super().__init__(**kwargs)
 line 92, in __init__ self.update_data()
line 76, in update_data  self.ele_shapes = layout.create_element_entities(self)
line 535, in create_element_entities if opm['ss'].conjugate_type == 'infinite':
AttributeError: 'NoneType' object has no attribute 'conjugate_type'`

From the previous question I added the `sm.set_stop()` and I'm trying to control the stop size with `sm.stop_surface = 40`.  I've also deviated from the example by trying to control the model using the `f/#` as ultimately I'm wanting two rays.

**Comment by mjhoptics** [comment, marked as answer]
Hi @cbhirsch,
I've found a couple of things.

1. The stop_surface attribute in seq_model is an index into the ifcs list, not a specification of the stop size. Commenting this out stops the error output.
2. The set_stop() method should be used after the stop surface is added. So moving it after the first add_surface invocation will achieve what you want here. (By default, the first surface is used if no stop specified, so not super critical *in this case*).
3. The aperture size is controlled by the pupil specification, in this case the image space f/#.
4. To create a biconvex lens, the radius of the second surface should be negative, i.e. the surface center of curvature is to the left of this surface.

Here's my code:
```
isdark = False

# use standard rayoptics environment
from rayoptics.environment import *

# util functions
from rayoptics.util.misc_math import normalize

#Creating a new, empty, model
opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']

#units
opm.system_spec.dimensions = 'mm'

#Specify aperture, field, and wavelengths
osp['pupil'] = PupilSpec(osp, key=['image', 'f/#'], value=10)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=5, flds=[0. , 1.], is_relative=True)
osp['wvls'] = WvlSpec([('F', 0.5), (587.5618, 1.0), ('C', 0.5)], ref_wl=1)
#sm.stop_surface = 40

opm.radius_mode = True

sm.gaps[0].thi=1e10

sm.add_surface([206, 1, 'N-BK7', 'Schott'])
sm.set_stop()
sm.add_surface([-206, 5.86])

opm.update_model()
sm.list_model()
```
Regards, Mike

**Reply by cbhirsch** [reply]
Thanks for the help this worked
